website/static/js/pages/FileEditor.js

```diff
@@ -0,0 +1,176 @@
+'use strict';
+var ko = require('knockout');
+var $ = require('jquery');
+var $osf = require('js/osfHelpers');
+var Raven = require('raven-js');
+var Markdown = require('pagedown-ace-converter');
+Markdown.getSanitizingConverter = require('pagedown-ace-sanitizer').getSanitizingConverter;
+require('imports?Markdown=pagedown-ace-converter!pagedown-ace-editor');
+
+
+/**
+ * Binding handler that instantiates an ACE editor.
+ * The value accessor must be a ko.observable.
+ * Example: <div data-bind="ace: currentText" id="editor"></div>
+ */
+var editor;
+ko.bindingHandlers.ace = {
+    init: function (element, valueAccessor) {
+        editor = ace.edit(element.id);  // jshint ignore: line
+
+        // Updates the view model based on changes to the editor
+        editor.getSession().on('change', function () {
+            valueAccessor()(editor.getValue());
+        });
+    },
+    update: function (element, valueAccessor) {
+        var content = editor.getValue();        // Content of ace editor
+        var value = ko.unwrap(valueAccessor()); // Value from view model
+
+        // Updates the editor based on changes to the view model
+        if (value !== undefined && content !== value) {
+            var cursorPosition = editor.getCursorPosition();
+            editor.setValue(value);
+            editor.gotoLine(cursorPosition.row + 1, cursorPosition.column);
+        }
+    }
+};
+
+function ViewModel(url, viewText) {
+    var self = this;
+
+    self.initText = ko.observable('');
+    self.currentText = viewText; //from filePage's VM
+    self.activeUsers = ko.observableArray([]);
+    self.status = ko.observable('connecting');
+    self.throttledStatus = ko.observable(self.status());
+
+    self.displayCollaborators = ko.computed(function() {
+       return self.activeUsers().length > 1;
+    });
+
+    // Throttle the display when updating status.
+    self.updateStatus = function() {
+        self.throttledStatus(self.status());
+    };
+
+    self.throttledUpdateStatus = $osf.throttle(self.updateStatus, 4000, {leading: false});
+
+    self.status.subscribe(function (newValue) {
+        if (newValue !== 'connecting') {
+            self.updateStatus();
+        }
+
+        self.throttledUpdateStatus();
+    });
+
+    self.statusDisplay = ko.computed(function() {
+        switch(self.throttledStatus()) {
+            case 'connected':
+                return 'Live editing mode';
+            case 'connecting':
+                return 'Attempting to connect';
+            case 'unsupported':
+                return 'Unsupported browser';
+            default:
+                return 'Unavailable: Live editing';
+        }
+    });
+
+    self.progressBar = ko.computed(function() {
+        switch(self.throttledStatus()) {
+            case 'connected':
```

Comment:
It might be cleaner to do:
```javascript
var className = 'progress-bar ';
 switch(self.throttledStatus()) {
            case 'connected':
                 className += 'progress-bar-success';
            case 'connecting':
                  className += 'progress-bar-warning progress-bar-striped active';
            default:
                 className += 'progress-bar-danger';
        }
return {
    class: className
}
```